Add sample DAG, test ExasolToS3Operator, and update documentation (#22632)

jieyao-MilestoneHub opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements. See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership. The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License. You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied. See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

### Summary
Adds documentation for ExasolToS3Operator to enhance clarity and usability.

### Changes Introduced
- **New Documentation**: exasol_to_s3.rst (Setup guide & example usage)

### Related Issues
related: #22632

Commit:
Ready to do PR

vincbeck reviewed (changes requested):
Those files are empty

Commit:
docs: Add Exasol connection setup guide for Airflow

This commit adds a detailed Exasol connection setup guide for Apache Airflow, outlining the configuration steps, prerequisites, and troubleshooting tips. The document is structured in reStructuredText (.rst) format for better readability and integration with documentation tools.

Requires review to ensure clarity and completeness.

Commit:
feat(exasol): Optimized Exasol connection and improve documentation

- Optimized `ExasolToS3Operator` to separate Exasol and S3 operations.
- Improved unit tests for `ExasolExportOperator`, `S3UploadOperator`, and `ExasolToS3Operator`.
- Added `example_exasol_to_s3` DAG to validate Exasol connection.
- Created `exasol_connection.rst` with setup instructions, avoiding credit card requirement.

vincbeck commented on airflow/example_dags/example_exasol_to_s3.py:
```diff
@@ -0,0 +1,53 @@
+    tags=["exasol", "test"],
+) as dag:
+
+    test_exasol_connection = ExasolExportOperator(
```
`test_exasol_connection` does not describe well what this operator does

jieyao-MilestoneHub replied:
Got it, I will rewrite `airflow/example_dags/example_exasol_to_s3.py` to focus on Exasol connection, S3 query, and export to S3.

vincbeck commented on airflow/example_dags/example_exasol_to_s3.py:
```diff
@@ -0,0 +1,53 @@
+This DAG uses the ExasolToS3Operator to execute a simple SQL query on Exasol
+and simulate an S3 upload. The main purpose is to verify that the Exasol connection
+settings (exasol_default) are working correctly. S3 connection is simulated via dummy
+parameters (aws_default, dummy bucket and key).
```
An example DAG is a DAG that should work out of the box (if you have valid credentials of course). I would rephrase this comment and just mention what this DAG is doing. This DAG should work for someone who has exasol and AWS permissions.

jieyao-MilestoneHub replied:
Got it, I misunderstood the purpose of the DAG—accidentally fell into the Exasol testing trap. I will modify it.

vincbeck commented on airflow/example_dags/exasol_connection.rst:
```diff
@@ -0,0 +1,152 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
```
There should no doc in `example_dags` folder. This kind of documentation should go in `providers/exasol/docs`

jieyao-MilestoneHub replied:
Got it.

vincbeck commented on providers/amazon/docs/connections/exasol.rst:
```diff
@@ -0,0 +1,174 @@
+.. _exasol_airflow_setup:
```
Please do not duplicate documentation. Plus, this documentation is related to Exasol, it has nothing to do with Amazon. As I mentioned in my previous comment, this should go in Exasol documentation

vincbeck commented on providers/amazon/src/airflow/providers/amazon/aws/transfers/exasol_to_s3.py:
```diff
@@ -31,29 +32,165 @@
     from airflow.utils.context import Context
 
 
+class ExasolExportOperator(BaseOperator):
```
We already have `ExasolToS3Operator`, are you sure we need this operator as well?

jieyao-MilestoneHub replied:
I initially thought example_exasol_to_s3 should only test the Exasol connection, so it should work without S3 credentials. This was the main reason for my changes.

Regarding the code modifications, I first planned to outline my approach (as I mentioned focusing on separation of concerns) before discussing further. As a result, my changes were somewhat disorganized—sorry for any confusion caused.

Now that I’ve clarified the issue, I will make further adjustments:

1. Revert exasol_to_s3.py and test_exasol_to_s3.py to their original state.
2. Add example_exasol_to_s3.py to test the full functionality (Exasol connection, S3 query, and export to S3).

vincbeck commented on providers/amazon/src/airflow/providers/amazon/aws/transfers/exasol_to_s3.py:
```diff
@@ -31,29 +32,165 @@
+            raise
+
+
+class S3UploadOperator(BaseOperator):
```
`S3CreateObjectOperator` already exists

jieyao-MilestoneHub replied:
Replied in the previous comment.

vincbeck commented on providers/amazon/src/airflow/providers/amazon/aws/transfers/exasol_to_s3.py:
```diff
@@ -86,27 +223,28 @@ def __init__(
-    def execute(self, context: Context):
-        exasol_hook = ExasolHook(exasol_conn_id=self.exasol_conn_id)
-        s3_hook = S3Hook(aws_conn_id=self.aws_conn_id)
+    def execute(self, context: Context) -> str:
```
Why modifying the implementation? Did you notice anything wrong with the previous implementation?

jieyao-MilestoneHub replied:
Replied to the comment before the last one.

Commit:
feat: Added exasol connection test dag and guided document

- Added example_exasol_to_s3 following the rst rule
- Added exasol_ce_setup to describe detail to connection
- Added exsol to describe standard connection info

jieyao-MilestoneHub commented on providers/amazon/tests/unit/amazon/aws/transfers/test_exasol_to_s3.py:
IRestore to its original state

jieyao-MilestoneHub commented on providers/amazon/src/airflow/providers/amazon/aws/transfers/exasol_to_s3.py:
Restore to its original state

vincbeck replied:
`return self.key` is still added from the original version. Is it intended?

jieyao-MilestoneHub replied:
Thanks for pointing that out! That was an oversight on my part — it's not intended to be there. I'll remove it in the next update.

vincbeck commented on providers/amazon/tests/system/amazon/aws/example_exasol_to_s3.py:
```diff
@@ -0,0 +1,91 @@
+        aws_conn_id="aws_default",
+        export_params={},
+        query_params={},
+        use_s3=True,
```
This parameter does not exist?

```suggestion
```

jieyao-MilestoneHub replied:
I overlooked this change last time. I’ll fix it in this update.

vincbeck commented on providers/amazon/tests/system/amazon/aws/example_exasol_to_s3.py:
```diff
@@ -0,0 +1,91 @@
+        exasol_conn_id="exasol_default",
+        aws_conn_id="aws_default",
+        export_params={},
+        query_params={},
```
nit: let's try to keep the example DAG as minimal as possible. These are default values, so there is no need to define them.

```suggestion
```

jieyao-MilestoneHub replied:
Got it! Thanks for the reminder! I’ll update it in the next change.

vincbeck commented on providers/amazon/tests/system/amazon/aws/example_exasol_to_s3.py:
```diff
@@ -0,0 +1,91 @@
+        trigger_rule=TriggerRule.ALL_DONE,
+    )
+
+    create_s3_bucket >> exasol_to_s3 >> delete_s3_bucket
```
nit: in Amazon provider package we are trying to be consistent and use `chain` instead.

```suggestion
    chain(test_context, create_s3_bucket, exasol_to_s3, delete_s3_bucket)
```

jieyao-MilestoneHub replied:
Got it! Thanks for the reminder! I’ll update it in the next change.

vincbeck commented on providers/amazon/tests/unit/amazon/aws/operators/test_exasol_to_s3.py:
Please add unit tests

jieyao-MilestoneHub replied:
Is this for the unit test of `example_exasol_to_s3.py`? I didn’t find unit tests for other `example_XXX_to_s3.py` files.

vincbeck replied:
You do not modify the operator  `ExasolToS3Operator` so you do not need to create this file. My point is, you either create a file and add unit tests in them, or you do not create a file

jieyao-MilestoneHub replied:
I understand. Thanks for the explanation! I won’t add any new files in this update.

vincbeck reviewed (commented):
Thanks for the update! This is way better. I added some comments but overall it looks good

jieyao-MilestoneHub reviewed (commented):
Didn’t realize comments could be 'pending'—thought I had already sent them. Thanks for your patience!

Commit:
Simplify example DAG and improve consistency with Amazon provider standards

- Removed default parameters (exasol_conn_id, aws_conn_id, export_params, query_params) to keep the example DAG minimal and focused
- Removed non-existent parameter `use_s3` from ExasolToS3Operator
- Replaced task chaining with `chain()` function for consistency across Amazon provider examples

vincbeck commented on providers/amazon/tests/unit/amazon/aws/operators/test_exasol_to_s3.py:
Please remove this file

jieyao-MilestoneHub replied:
Hi! Just to clarify — this test file already exists under `providers/amazon/tests/unit/amazon/aws/operators/` and is consistent with other `test_*_to_s3.py` operator tests in the same directory. I'm only updating the existing file, not introducing a new one.

As far as I can see, this is the only location in the repo where all `*_to_s3.py` tests are placed, so I'm not aware of a newer or preferred test directory structure. Please let me know if there's an official plan to move them elsewhere — otherwise I suggest we keep the file here for consistency.
